File: backend/routers/staff.py
# Staff management router for Silent Honor Foundation
import asyncio
import secrets
from datetime import datetime, timezone
from fastapi import APIRouter, HTTPException, Request
from bson import ObjectId

from middleware.auth_middleware import get_current_admin
from middleware.logging_middleware import log_audit_event, AUDIT_ACTIONS
from utils.auth import hash_password
from utils.validators import StaffRequest
from utils.email import send_staff_welcome_email

STAFF_ROLES = ["staff", "admin", "counselor"]

router = APIRouter(prefix="/api/admin/staff", tags=["Staff"])

# Database reference
db = None

def set_db(database):
    global db
    db = database

def generate_temp_password():
    return secrets.token_urlsafe(12)

@router.get("")
async def get_staff(request: Request):
    await get_current_admin(request)
    staff = await db.users.find({"role": {"$in": ["staff", "admin", "counselor"]}}).to_list(100)
    result = []
    for s in staff:
        client_count = 0
        if s.get("role") == "counselor":
            client_count = await db.users.count_documents({"assigned_counselor_id": s["_id"]})
        result.append({
            "id": str(s["_id"]),
            "email": s.get("email"),
            "first_name": s.get("first_name", ""),
            "last_name": s.get("last_name", ""),
            "name": f"{s.get('first_name', '')} {s.get('last_name', '')}".strip(),
            "role": s.get("role"),
            "title": s.get("title", ""),
            "bio": s.get("bio", ""),
            "specialties": s.get("specialties", []),
            "permissions": s.get("permissions", []),
            "active": s.get("active", True),
            "client_count": client_count,
            "created_at": s.get("created_at").isoformat() if s.get("created_at") else None,
            "last_active": s.get("last_active").isoformat() if s.get("last_active") else None
        })
    return result

@router.post("")
async def create_staff(request: Request, data: StaffRequest):
    admin = await get_current_admin(request)

    if data.role not in STAFF_ROLES:
        raise HTTPException(status_code=400, detail=f"Role must be one of: {STAFF_ROLES}")

    existing = await db.users.find_one({"email": data.email.lower()})
    if existing:
        raise HTTPException(status_code=400, detail="Email already registered")

    temp_password = generate_temp_password()

    user_doc = {
        "email": data.email.lower(),
        "password_hash": hash_password(temp_password),
        "first_name": data.first_name,
        "last_name": data.last_name,
        "role": data.role,
        "title": data.title or "",
        "permissions": data.permissions or [],
        "active": True,
        "created_at": datetime.now(timezone.utc)
    }

    result = await db.users.insert_one(user_doc)
    user_id = str(result.inserted_id)

    await log_audit_event(
        action=AUDIT_ACTIONS.get("STAFF_CREATED", "STAFF_CREATED"),
        entity_type="user",
        entity_id=user_id,
        user_email=admin.get("email"),
        details={"role": data.role, "new_staff_email": data.email}
    )

    asyncio.create_task(send_staff_welcome_email(data.email, data.first_name, data.role, temp_password))

    return {
        "id": user_id,
        "email": data.email,
        "first_name": data.first_name,
        "last_name": data.last_name,
        "name": f"{data.first_name} {data.last_name}".strip(),
        "role": data.role,
        "active": True,
        "created_at": datetime.now(timezone.utc).isoformat()
    }

@router.put("/{staff_id}")
async def update_staff(request: Request, staff_id: str):
    admin = await get_current_admin(request)
    data = await request.json()

    allowed = ["first_name", "last_name", "email", "role", "title", "active"]
    update_data = {k: v for k, v in data.items() if k in allowed}

    if "role" in update_data and update_data["role"] not in STAFF_ROLES:
        raise HTTPException(status_code=400, detail=f"Role must be one of: {STAFF_ROLES}")

    if not update_data:
        raise HTTPException(status_code=400, detail="No valid fields to update")

    update_data["updated_at"] = datetime.now(timezone.utc)

    result = await db.users.update_one(
        {"_id": ObjectId(staff_id), "role": {"$in": STAFF_ROLES}},
        {"$set": update_data}
    )

    if result.matched_count == 0:
        raise HTTPException(status_code=404, detail="Staff member not found")

    await log_audit_event(
        action=AUDIT_ACTIONS.get("STAFF_UPDATED", "STAFF_UPDATED"),
        entity_type="user",
        entity_id=staff_id,
        user_email=admin.get("email"),
        details=update_data
    )

    return {"message": "Staff member updated"}

@router.get("/counselors")
async def get_counselors(request: Request):
    await get_current_admin(request)
    counselors = await db.users.find({"role": "counselor", "active": True}).to_list(100)
    result = []
    for c in counselors:
        client_count = await db.users.count_documents({"assigned_counselor_id": c["_id"]})
        result.append({
            "id": str(c["_id"]),
            "name": f"{c.get('first_name', '')} {c.get('last_name', '')}".strip(),
            "email": c.get("email"),
            "specialties": c.get("specialties", []),
            "client_count": client_count
        })
    return result

File: backend-rust/crates/api/src/routes/staff.rs
//! Staff management (`/api/admin/staff/*`). Port of routers/staff.py. Staff,
//! admins and counselors are all `users` documents with the matching `role`.

use axum::extract::{Path, State};
use axum::http::HeaderMap;
use axum::Json;
use bson::oid::ObjectId;
use bson::{doc, Bson, Document};
use serde::Deserialize;
use serde_json::{json, Value};

use super::{authenticate_admin, ddate, dstr, dstr_or, full_name, hex_id, iso, log_audit};
use crate::state::AppState;
use sh_core::auth::hash_password;
use sh_core::{AppError, AppResult};

const STAFF_ROLES: [&str; 3] = ["staff", "admin", "counselor"];

fn rand_hex(n: usize) -> String {
    let mut buf = vec![0u8; n];
    getrandom::getrandom(&mut buf).ok();
    buf.iter().map(|b| format!("{b:02x}")).collect()
}

fn arr(d: &Document, k: &str) -> Value {
    match d.get_array(k) {
        Ok(a) => Value::Array(a.iter().cloned().map(super::sanitize).collect()),
        Err(_) => json!([]),
    }
}

// ── GET /api/admin/staff ──────────────────────────────────────────────────────
pub async fn get_staff(State(state): State<AppState>, headers: HeaderMap) -> AppResult<Json<Value>> {
    authenticate_admin(&state, &headers).await?;
    let users = state.db.collection::<Document>("users");
    let mut cur = users
        .find(doc! { "role": { "$in": ["staff", "admin", "counselor"] } })
        .limit(100)
        .await?;
    let mut out: Vec<Value> = Vec::new();
    while cur.advance().await? {
        let s = cur.deserialize_current()?;
        let client_count = if s.get_str("role").unwrap_or("") == "counselor" {
            if let Ok(sid) = s.get_object_id("_id") {
                users.count_documents(doc! { "assigned_counselor_id": sid }).await?
            } else {
                0
            }
        } else {
            0
        };
        out.push(json!({
            "id": hex_id(&s),
            "email": dstr(&s, "email"),
            "first_name": dstr_or(&s, "first_name", ""),
            "last_name": dstr_or(&s, "last_name", ""),
            "name": full_name(&s),
            "role": dstr(&s, "role"),
            "title": dstr_or(&s, "title", ""),
            "bio": dstr_or(&s, "bio", ""),
            "specialties": arr(&s, "specialties"),
            "permissions": arr(&s, "permissions"),
            "active": Value::Bool(s.get_bool("active").unwrap_or(true)),
            "client_count": client_count,
            "calendly_url": dstr_or(&s, "calendly_url", ""),
            "created_at": ddate(&s, "created_at"),
            "last_active": ddate(&s, "last_active"),
        }));
    }
    Ok(Json(json!(out)))
}

#[derive(Debug, Deserialize)]
pub struct StaffRequest {
    pub email: String,
    #[serde(default)]
    pub first_name: String,
    #[serde(default)]
    pub last_name: String,
    #[serde(default = "staff_role")]
    pub role: String,
    #[serde(default)]
    pub title: Option<String>,
    #[serde(default)]
    pub bio: Option<String>,
    #[serde(default)]
    pub specialties: Vec<String>,
    #[serde(default)]
    pub credentials: Option<String>,
    #[serde(default)]
    pub permissions: Vec<String>,
    #[serde(default)]
    pub calendly_url: Option<String>,
}
fn staff_role() -> String {
    "staff".to_string()
}

// ── POST /api/admin/staff ─────────────────────────────────────────────────────
pub async fn create_staff(
    State(state): State<AppState>,
    headers: HeaderMap,
    Json(body): Json<StaffRequest>,
) -> AppResult<Json<Value>> {
    let (_, admin) = authenticate_admin(&state, &headers).await?;
    if !STAFF_ROLES.contains(&body.role.as_str()) {
        return Err(AppError::BadRequest(
            "Role must be one of: [staff, admin, counselor]".to_string(),
        ));
    }
    let email = body.email.to_lowercase();
    let users = state.db.collection::<Document>("users");
    if users.find_one(doc! { "email": email.as_str() }).await?.is_some() {
        return Err(AppError::BadRequest("Email already registered".to_string()));
    }

    let temp_password = rand_hex(12);
    let hash = hash_password(&temp_password).map_err(AppError::Internal)?;
    let now = bson::DateTime::now();
    let specialties: Vec<Bson> = body.specialties.iter().map(|s| Bson::String(s.clone())).collect();
    let permissions: Vec<Bson> = body.permissions.iter().map(|s| Bson::String(s.clone())).collect();

    let res = users
        .insert_one(doc! {
            "email": email.as_str(),
            "password_hash": hash.as_str(),
            "first_name": body.first_name.as_str(),
            "last_name": body.last_name.as_str(),
            "role": body.role.as_str(),
            "roles": [body.role.as_str()],
            "title": body.title.clone().unwrap_or_default(),
            "bio": body.bio.clone().unwrap_or_default(),
            "specialties": specialties,
            "credentials": body.credentials.clone().unwrap_or_default(),
            "calendly_url": match &body.calendly_url { Some(u) => Bson::String(u.clone()), None => Bson::Null },
            "permissions": permissions,
            "active": true,
            "created_at": now,
        })
        .await?;
    let id = res.inserted_id.as_object_id().map(|o| o.to_hex()).unwrap_or_default();

    log_audit(&state, "staff_created", "user", Some(&id), Some(admin.email.as_str())).await;
    // Awaited (Lambda freezes background tasks on return).
    sh_core::email::send_staff_welcome_email(&email, &body.first_name, &body.role, &temp_password).await;

    Ok(Json(json!({
        "id": id,
        "email": body.email,
        "first_name": body.first_name,
        "last_name": body.last_name,
        "name": format!("{} {}", body.first_name, body.last_name).trim(),
        "role": body.role,
        "active": true,
        "created_at": iso(Some(now)),
    })))
}

// ── PUT /api/admin/staff/{staff_id} ───────────────────────────────────────────
pub async fn update_staff(
    State(state): State<AppState>,
    headers: HeaderMap,
    Path(staff_id): Path<String>,
    Json(body): Json<Value>,
) -> AppResult<Json<Value>> {
    let (_, admin) = authenticate_admin(&state, &headers).await?;
    let oid = ObjectId::parse_str(&staff_id).map_err(|_| AppError::NotFound)?;

    const ALLOWED: [&str; 11] = [
        "first_name", "last_name", "email", "role", "title", "active", "bio", "specialties",
        "calendly_url", "credentials", "max_caseload",
    ];
    let mut set = Document::new();
    if let Some(obj) = body.as_object() {
        for (k, v) in obj {
            if ALLOWED.contains(&k.as_str()) {
                set.insert(k, bson::to_bson(v).unwrap_or(Bson::Null));
            }
        }
    }
    if let Some(role) = set.get_str("role").ok() {
        if !STAFF_ROLES.contains(&role) {
            return Err(AppError::BadRequest(
                "Role must be one of: [staff, admin, counselor]".to_string(),
            ));
        }
    }
    if set.is_empty() {
        return Err(AppError::BadRequest("No valid fields to update".to_string()));
    }
    set.insert("updated_at", bson::DateTime::now());
    let res = state
        .db
        .collection::<Document>("users")
        .update_one(
            doc! { "_id": oid, "role": { "$in": ["staff", "admin", "counselor"] } },
            doc! { "$set": set },
        )
        .await?;
    if res.matched_count == 0 {
        return Err(AppError::NotFound);
    }
    log_audit(&state, "staff_updated", "user", Some(&staff_id), Some(admin.email.as_str())).await;
    Ok(Json(json!({ "message": "Staff member updated" })))
}

// ── GET /api/admin/staff/counselors ───────────────────────────────────────────
pub async fn get_counselors(
    State(state): State<AppState>,
    headers: HeaderMap,
) -> AppResult<Json<Value>> {
    authenticate_admin(&state, &headers).await?;
    let users = state.db.collection::<Document>("users");
    let mut cur = users
        .find(doc! { "role": "counselor", "active": true })
        .limit(100)
        .await?;
    let mut out: Vec<Value> = Vec::new();
    while cur.advance().await? {
        let c = cur.deserialize_current()?;
        let client_count = match c.get_object_id("_id") {
            Ok(cid) => users.count_documents(doc! { "assigned_counselor_id": cid }).await?,
            Err(_) => 0,
        };
        out.push(json!({
            "id": hex_id(&c),
            "name": full_name(&c),
            "email": dstr(&c, "email"),
            "specialties": arr(&c, "specialties"),
            "client_count": client_count,
        }));
    }
    Ok(Json(json!(out)))
}
